fsm.go

```diff
@@ -184,16 +163,21 @@ func (fsm *FSM) updateSubscribers() {
 
 // fsmSnapshot implements the hashicorp/raft interface and allows to serialize a
 // state into a byte slice that can be used as a snapshot of the system.
-type fsmSnapshot []byte
+type fsmSnapshot struct {
+	fsm *FSM
+}
 
-// Persist writes the snapshot (a serialized state) to a raft.SnapshotSink
-func (fsms fsmSnapshot) Persist(sink raft.SnapshotSink) error {
-	_, err := sink.Write(fsms)
+// Persist writes the snapshot (a serialized state) to a raft.SnapshotSink.
+func (snap *fsmSnapshot) Persist(sink raft.SnapshotSink) error {
+	snap.fsm.mux.Lock()
+	defer snap.fsm.mux.Unlock()
+	err := EncodeSnapshot(snap.fsm.state, sink)
 	if err != nil {
+		sink.Cancel()
 		return err
 	}
-	return sink.Close()
+	sink.Close()
```

Comment:
Why not return any potential error from `sink.Close()`?

Reply:
hmm I have no idea